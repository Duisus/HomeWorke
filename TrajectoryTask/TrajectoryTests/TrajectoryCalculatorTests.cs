using System.Drawing;
using System.Linq;
using ApprovalTests;
using ApprovalTests.Reporters;
using FluentAssertions;
using NUnit.Framework;
using Trajectory;

namespace TrajectoryTests
{
    public class TrajectoryCalculatorTests
    {
        private const float G = 9.81f;
        private TrajectoryCalculator calculator;

        [SetUp]
        public void Setup()
        {
            calculator = new TrajectoryCalculator(
                new PointF(0, 0),
                50,
                30,
                1.2f,
                0.4f);
        }

        [Test]
        public void GetPoints_FirstPointIsStartPointFromCtor()
        {
            var startPoint = new PointF(10.3f, 15);
            calculator = new TrajectoryCalculator(
                startPoint, 40.31f, 30, 3.1f, 1);

            calculator.GetPoints(0.1f).First().Coords
                .Should().Be(startPoint);
        }

        [Test]
        public void GetPoints_LastPointHasZeroYCoordinate()
        {
            calculator.GetPoints(0.1f).Last().Coords.Y
                .Should().Be(0);
        }

        [Test]
        public void GetPoint_LastPointHasXCoordinateThatEqualToFlightRange()
        {
            calculator.GetPoints(0.01f).Last().Coords.X
                .Should().BeApproximately(5, 0.05f); // TODO refactor
        }

        [Test]
        public void GetPoint_HighestPointHasCorrectYCoordinate()
        {
            calculator.GetPoints(0.01f).Max(trajPoint => trajPoint.Coords.Y)
                .Should().BeApproximately(1.75f, 0.05f); // TODO refactor
        }

        [Test]
        [UseReporter(typeof(VisualStudioReporter))]
        public void ApprovalTest()
        {
            var resultPoints = calculator.GetPoints(0.1f);
            var strFromAllPoints = string.Join(
                '\n', resultPoints.Select(p => p.ToString()));
            
            Approvals.Verify(strFromAllPoints);
        }
    }
}