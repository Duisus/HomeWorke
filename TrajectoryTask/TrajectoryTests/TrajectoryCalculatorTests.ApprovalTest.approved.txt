Time: 0, Coords: {X=0, Y=0}, Speed: (43,30127, 25)
Time: 0,1, Coords: {X=4,3301, Y=2,5}, Speed: (41,85789, 23,185667)
Time: 0,2, Coords: {X=8,5159, Y=4,8186}, Speed: (40,462627, 21,431812)
Time: 0,3, Coords: {X=12,5622, Y=6,9618}, Speed: (39,113873, 19,736418)
Time: 0,4, Coords: {X=16,4736, Y=8,9354}, Speed: (37,810074, 18,097538)
Time: 0,5, Coords: {X=20,2546, Y=10,7452}, Speed: (36,549736, 16,513287)
Time: 0,6, Coords: {X=23,9096, Y=12,3965}, Speed: (35,33141, 14,981844)
Time: 0,70000005, Coords: {X=27,4427, Y=13,8947}, Speed: (34,153694, 13,50145)
Time: 0,8000001, Coords: {X=30,8581, Y=15,2448}, Speed: (33,015236, 12,070401)
Time: 0,9000001, Coords: {X=34,1596, Y=16,4518}, Speed: (31,914726, 10,687055)
Time: 1,0000001, Coords: {X=37,3511, Y=17,5205}, Speed: (30,8509, 9,349819)
Time: 1,1000001, Coords: {X=40,4362, Y=18,4555}, Speed: (29,822536, 8,057158)
Time: 1,2000002, Coords: {X=43,4185, Y=19,2612}, Speed: (28,828451, 6,8075867)
Time: 1,3000002, Coords: {X=46,3013, Y=19,942}, Speed: (27,867502, 5,599667)
Time: 1,4000002, Coords: {X=49,0881, Y=20,502}, Speed: (26,938585, 4,4320116)
Time: 1,5000002, Coords: {X=51,782, Y=20,9452}, Speed: (26,040632, 3,303278)
Time: 1,6000003, Coords: {X=54,3861, Y=21,2755}, Speed: (25,172611, 2,2121687)
Time: 1,7000003, Coords: {X=56,9034, Y=21,4967}, Speed: (24,333523, 1,1574297)
Time: 1,8000003, Coords: {X=59,3368, Y=21,6124}, Speed: (23,522406, 0,13784862)
Time: 1,9000003, Coords: {X=61,689, Y=21,6262}, Speed: (22,738325, -0,84774643)
Time: 2,0000002, Coords: {X=63,9628, Y=21,5414}, Speed: (21,980381, -1,8004882)
Time: 2,1000001, Coords: {X=66,1608, Y=21,3614}, Speed: (21,247702, -2,721472)
Time: 2,2, Coords: {X=68,2856, Y=21,0893}, Speed: (20,539444, -3,6117563)
Time: 2,3, Coords: {X=70,3395, Y=20,7281}, Speed: (19,854795, -4,4723644)
Time: 2,3999999, Coords: {X=72,325, Y=20,2809}, Speed: (19,192968, -5,3042855)
Time: 2,4999998, Coords: {X=74,2443, Y=19,7505}, Speed: (18,553202, -6,108476)
Time: 2,5999997, Coords: {X=76,0996, Y=19,1397}, Speed: (17,934761, -6,8858604)
Time: 2,6999996, Coords: {X=77,8931, Y=18,4511}, Speed: (17,336935, -7,637332)
Time: 2,7999995, Coords: {X=79,6268, Y=17,6874}, Speed: (16,759037, -8,363754)
Time: 2,8999994, Coords: {X=81,3027, Y=16,851}, Speed: (16,200401, -9,065963)
Time: 2,9999993, Coords: {X=82,9227, Y=15,9444}, Speed: (15,660387, -9,744764)
Time: 3,0999992, Coords: {X=84,4887, Y=14,9699}, Speed: (15,138373, -10,400939)
Time: 3,199999, Coords: {X=86,0025, Y=13,9298}, Speed: (14,63376, -11,035241)
Time: 3,299999, Coords: {X=87,4659, Y=12,8263}, Speed: (14,145968, -11,648399)
Time: 3,399999, Coords: {X=88,8805, Y=11,6615}, Speed: (13,674436, -12,241119)
Time: 3,4999988, Coords: {X=90,2479, Y=10,4374}, Speed: (13,21862, -12,814082)
Time: 3,5999987, Coords: {X=91,5698, Y=9,156}, Speed: (12,777999, -13,367947)
Time: 3,6999986, Coords: {X=92,8476, Y=7,8192}, Speed: (12,352065, -13,903349)
Time: 3,7999985, Coords: {X=94,0828, Y=6,4289}, Speed: (11,94033, -14,420904)
Time: 3,8999984, Coords: {X=95,2768, Y=4,9868}, Speed: (11,542318, -14,921207)
Time: 3,9999983, Coords: {X=96,431, Y=3,4947}, Speed: (11,157574, -15,404834)
Time: 4,0999985, Coords: {X=97,5468, Y=1,9542}, Speed: (10,785654, -15,872339)
Time: 4,1999984, Coords: {X=98,6254, Y=0,367}, Speed: (10,426132, -16,32426)
Time: 4,2999983, Coords: {X=99,668, Y=0}, Speed: (10,078594, -16,761118)